### DISCUSSION TITLE
`dependency_overrides` not working through Path Operation decorator?

### DISCUSSION THREAD

**Discussion by nessadc** [fastapi/fastapi#6479] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends

from project.config import Settings


def get_settings():
    return Settings()


app = FastAPI()


@app.get("/info")
async def info(settings: Settings = Depends(get_settings)):
    print(settings)
    return {
        "app_name": settings.app_name,
        "admin_email": settings.admin_email,
        "items_per_user": settings.items_per_user,
    }


@app.post("/things", status_code=201, dependencies=[Depends(get_settings)])
async def update(email):
    print(get_settings())
    returned_thing = put_decision(email, get_settings())
    return {"thing": returned_thing}


def put_decision(email, settings):
    s = email
    s += " " + str(settings.items_per_user)
    return s
```

### Description

I am trying to run some tests with pytest against FastAPI. I am having an issue overriding a dependency that I've specified using `depends` and `dependency_overrides`. The `settings` get overridden fine in the GET endpoint but not in the POST endpoint. 

Here is my `main.py` file:
```python
from fastapi import FastAPI, Depends

from project.config import Settings


def get_settings():
    return Settings()


app = FastAPI()


@app.get("/info")
async def info(settings: Settings = Depends(get_settings)):
    print(settings)
    return {
        "app_name": settings.app_name,
        "admin_email": settings.admin_email,
        "items_per_user": settings.items_per_user,
    }


@app.post("/things", status_code=201, dependencies=[Depends(get_settings)])
async def update(email):
    print(get_settings())
    returned_thing = put_decision(email, get_settings())
    return {"thing": returned_thing}


def put_decision(email, settings):
    s = email
    s += " " + str(settings.items_per_user)
    return s
```

Here is my `test_ping.py` file
```python
from starlette.testclient import TestClient

from project.main import app, get_settings
from project.config import Settings


def get_settings_override():
    print("overriding!")

    return Settings(
        app_name="My cool API", admin_email="admin@example.com", items_per_user=30
    )


app.dependency_overrides[get_settings] = get_settings_override
client = TestClient(app)


def test_info():
    response = client.get("info")
    assert response.status_code == 200
    assert response.json() == {
        "admin_email": "admin@example.com",
        "app_name": "My cool API",
        "items_per_user": 30,
    }


def test_update():
    response = client.post("things", params={"email": "mything"})
    assert response.status_code == 201
    assert response.json() == {"thing": "mything 30"}
```

Here is my `config.py` file where the `Settings` class is first defined:
```python
from pydantic import BaseSettings


class Settings(BaseSettings):
    app_name: str = "My cool API"
    admin_email: str = "example@example.com"
    items_per_user: int = 50

```

What am I missing here? I've tried overriding the function directly by doing the below but doesn't seem like it makes a difference.
```python
app.get_settings = get_settings_override
```

I've looked at the solutions provided [here](https://github.com/tiangolo/fastapi/issues/641) and [here](https://github.com/tiangolo/fastapi/issues/656) but I am still observing the same behavior (i.e. my Settings stay the default for the POST call) but  

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

Python 3.9.8

**Comment by harunyasar** [comment]
@nessadc I don't think that you could override a dependency in the path operation decorator. You could simply patch the `get_settings` object directly.

```python
from unittest.mock import patch

from project import main

@patch.object(main, "get_settings", get_settings_override)
def test_update():
    response = client.post("things", params={"email": "mything"})
    assert response.status_code == 201
    assert response.json() == {"thing": "mything 30"}
```

I hope that it passes.

**Comment by nessadc** [comment]
@harunyasar I realized this as well after reading through some other issues related to `Depends`. Thank you for looking into this! This works for me now.

```python
import pytest
from starlette.testclient import TestClient

from . import main
from project.config import Settings


def get_settings_override():
    return Settings(
        app_name="My cool API", admin_email="admin@example.com", items_per_user=30
    )


@pytest.fixture(scope="module")
def test_app():
    main.app.dependency_overrides[main.get_settings] = get_settings_override
    with TestClient(main.app) as test_client:
        yield test_client


def test_info(test_app):
    response = test_app.get("info")
    assert response.status_code == 200
    assert response.json() == {
        "admin_email": "admin@example.com",
        "app_name": "My cool API",
        "items_per_user": 30,
    }


def test_update(test_app, monkeypatch):
    monkeypatch.setattr(main, "get_settings", get_settings_override)
    response = test_app.post("things", params={"email": "mything"})
    assert response.status_code == 201
    assert response.json() == {"thing": "mything 30"}

```

**Comment by YuriiMotov** [comment, marked as answer]
Just to clarify: when you use dependency function inside your endpoint function you are not using dependency injection, you are bypassing it.

```python
    returned_thing = put_decision(email, get_settings())  # Here you just uses get_settings function, bypassing dependency injection
```

Correct way is to include this dependency in function signature as you did for GET endpoint. This way you will use dependency injection full force and will not need monkeypatch
